### ISSUE TITLE
`task` code field in status

### ISSUE THREAD

**Issue by DavidGOrtega** [iterative/terraform-provider-iterative#357]
As stated here using two booleans may not cover all the desired states in the future. It would be better to subscribe to `status code` model.

To not break current implementation we could just add `code` and `code_desc`  fields within `status`

**Comment by 0x2b3bfa0** [comment]
Can you please add some examples of both the current and the desired status formats?

**Comment by 0x2b3bfa0** [comment]
🔔 @DavidGOrtega, and also an example use case, if there is a public repository?

**Comment by casperdcl** [comment]
I'm not sure what the underlying issue(s) are here.

- the categories? (ie. `running/succeeded/failed` isn't enough, we need `queued/error/terminated` etc?)
- how to implement the categories? (ie. are they computed from a bit field?)
- what should `parallelism` mean? How should users use it?

**Comment by 0x2b3bfa0** [comment]
The Y problem here is programmatically determining whether a task is ready to be disposed or not, with a short oneliner (presumably shorter than [this one](https://github.com/iterative/terraform-provider-iterative/issues/389#issuecomment-1074124839)).

The meaning of `parallelism` is completely out of scope for this issue, but it would be nice to open a new one and explain it properly. In a nutshell, it specifies the target size of the machine group; `parallelism = 3` will [try to] keep 3 [potentially spot] machines running until one of them finishes;[^1] all of them run exactly the same `script` and are [reachable] in the same [local] network; all of them have also a different public IP address.

[^1]: Purposeful design blunder: it should be “until three of them fail or succeed”, but that's a bit more complex than that and we should address that in the mid term.

**Comment by 0x2b3bfa0** [comment]
> _how to implement the categories? (ie. are they computed from a bit field?)_

Assuming that `running`, `succeeded` and `failed` are finite fields of order `parallelism + 1`, the optimal numeric representation would be as follows:

```python
>>> parallelism = 4
>>> running, succeeded, failed = 1, 2, 3
>>> status_code = sum(
...     (parallelism + 1)**exponent * value
...     for exponent, value in enumerate((running, succeeded, failed))
... )
```

Still, this is all but human–readable and, more importantly, status fields aren't guaranteed to be in range `[0, parallelism]`

**Comment by 0x2b3bfa0** [comment]
As per @DavidGOrtega's today conversation, he wanted the status code to be the script [exit status](https://en.wikipedia.org/wiki/Exit_status), but that would't play nicely with `parallelism` values other than 1. 🤨

**Comment by 0x2b3bfa0** [comment]
`<kidding>` Unless we define that code to be the float value resulting of averaging the exit status of each machine ([Solomonic](https://en.wikipedia.org/wiki/Judgement_of_Solomon) Solution). `</kidding>`

**Comment by 0x2b3bfa0** [comment]
As stated, I would suggest closing this issue in favor of:
1. Solving the CI/CD resource disposal use case with the one–liner at [#389 (comment)](https://github.com/iterative/terraform-provider-iterative/issues/389#issuecomment-1074124839)
2. Keeping the [`status`](https://registry.terraform.io/providers/iterative/iterative/latest/docs/resources/task#status) key–value map as an accurate description of the orchestrator status.
3. Keeping the [`logs`](https://registry.terraform.io/providers/iterative/iterative/latest/docs/resources/task#logs) list as the only source for script exit status information for every machine.
4. Documenting all of this?

**Comment by casperdcl** [comment]
What about using the `name_or_ID[:script_exit_code]`?

```
status = {
  running   = []
  succeeded = ['A', 'B', 'D', 'F']
  failed    = ['C:127', 'E:256']
}
```

**Comment by 0x2b3bfa0** [comment]
What is `name_or_ID`? 🤔

**Comment by casperdcl** [comment]
> What is `name_or_ID`? thinking

some unique per-parallel-task identifier.

**Comment by 0x2b3bfa0** [comment]
✈️

**Comment by casperdcl** [comment]
You don't want any per-parallel-task ID? Why? If not, could still do:

```
status = {
  running   = 0
  succeeded = 4
  failed    = ['script:127', 'docker:256', 'rclone:126']
}
```

**Comment by 0x2b3bfa0** [comment]
There is no way of persisting an identifier between spot instances for the same task.

1. Create a task with parallelism = 2.
2. Two spot instances appear and run the [same] script.
3. Both spot instances are terminated because the cloud is feeling adventurous.
4. Two new instances are created to replace those.

How can you be sure that those instances won't resume the same task (identifier for the case), instead of picking a different one?

Fixing that scenario requires some kind of leader election algorithm and a storage system with some desirable properties like atomicity and strict consistency.

**Comment by 0x2b3bfa0** [comment]
Test comment

**Comment by 0x2b3bfa0** [comment]
Test comment

**Comment by 0x2b3bfa0** [comment]
There is no way of persisting an identifier between spot instances for the same task.

1. Create a task with parallelism = 2.
2. Two spot instances appear and run the [same] script.
3. Both spot instances are terminated because the cloud is feeling adventurous.
4. Two new instances are created to replace those.

How can you be sure that those instances won't resume the same task (identifier for the case), instead of picking a different one?

Fixing that scenario requires some kind of leader election algorithm and a storage system with some desirable properties like atomicity and strict consistency.

**Comment by 0x2b3bfa0** [comment]
The only accessible and sensible enhancement I see would be something like this:

```
status = {
  running   = 0
  succeeded = 4
  failed    = [127, 13, 42]
}
```

**Comment by 0x2b3bfa0** [comment]
The question is: should we really change the schema of `failed` to represent exit codes? 🤔

**Comment by DavidGOrtega** [comment]
I would not even expose parallelism parameter right now. I find it totally useless.
What would be the scenario to use it?

**Comment by DavidGOrtega** [comment]
Indeed parallelism done correctly is not even a good term since we do not provide orchestrators. We we should define would be probably `actors` or `nodes` within a cluster is not more that merely interconnect  the nodes within the same cluster

**Comment by 0x2b3bfa0** [comment]
> _I would not even expose parallelism parameter right now. I find it totally useless._
> _What would be the scenario to use it?_

It's not as useful as [build matrix](https://docs.github.com/en/actions/using-jobs/using-a-build-matrix-for-your-jobs), but still can be useful if users bring their own task splitting system. See https://github.com/iterative/terraform-provider-iterative/issues/357#issuecomment-1076404436 for more information.

**Comment by DavidGOrtega** [comment]
> The current implementation of task uses the same parallelism concept as [CircleCI Parallelism](https://circleci.com/docs/2.0/parallelism-faster-jobs), sans the circleci split part.

Its not. CircleCI will provide one job to each node. CircleCI orchestrator (SAAS) will provide one job (executed in parallel) to every executor. Like GH does if you have multiple jobs within the yaml and you have multiple runners.

Right now as I say is not doing anything aside launching n instances to run strictly the same task

**Comment by 0x2b3bfa0** [comment]
> _Indeed parallelism done correctly is not even a good term since we do not provide orchestrators. The right parameter we would define probably is actors or nodes within a cluster. We should do not more that merely interconnect the nodes within the same cluster_

It sounds like a wording issue then (?)

**Comment by 0x2b3bfa0** [comment]
> _Conceptually we have the concept of ONE task and that task should have a status code @0x2b3bfa0 @casperdcl_

And what should that status code represent? 🤔

**Comment by 0x2b3bfa0** [comment]
> _Its not. CircleCI will provide one job to each node. CircleCI orchestrator (SAAS) will provide one job (executed in parallel) to every executor. Like GH does if you have multiple jobs within the yaml and you have multiple runners._

CircleCI runs the same code on every node, like we do. The difference is that it also provides a separate mechanism for splitting tests (files/lines/index) among nodes. 🤔 

> _Right now as I say is not doing anything aside launching n instances to run strictly the same task_

Indeed, although that's up to the user. Users can alter the behavior of their code based on some network– or storage–based task splitting mechanism.

**Comment by DavidGOrtega** [comment]
> It sounds like a wording issue then (?)

I would say no. It's conception. I might need to run multiple scripts (tasks) in parallel using the definition above and that would be parallelism. But I probably would need to run ONE task with multiple machines and here our provider only needs to run ONE script in one machine while the other are orchestrated by that single script with just one ONLY status code.

I think that the biggest concept misunderstanding that we have is that we have to logically define a script as a Task like other  ideas (GH jobs, CircleCI steps) etc...
One script is ONE Task and it should have ONE status code.
